activerecord/lib/active_record/autosave_association.rb

```diff
@@ -365,7 +365,12 @@ def normalize_reflection_attribute(indexed_attribute, reflection, index, attribu
       # Is used as a before_save callback to check while saving a collection
       # association whether or not the parent was a new record before saving.
       def before_save_collection_association
-        @new_record_before_save = new_record?
+        # If we defined a callback that updates the object before we defined the association, then this ivar may end up being
+        # manipulated to being `false` when it should be `true`. We guard this be only defining it once.
+        # See https://github.com/rails/rails/issues/38120 for more details
+        unless defined?(@new_record_before_save)
```

Comment:
Not sure if I get this fix. This instance variable is only set on associations callback so I don't see how `@new_record_before_save` would be defined when the `before_save_collection_association ` is called. Could you explain?

I can see how `new_record?` return `false` if a callback before this one salved the record, but if no other `before_save_collection_association` have run yet the `@new_record_before_save` should not be defined.

Reply:
The linked issue has more details, but I'll go into it a bit more here too :)

autosave associations works on a before_save. What I was experiencing is that in the case where a callback is defined before the has many (and that callback updates the model), then the children records would never get updated with the parent id.

This is because the following would happen:

1. Record create called
2. Autosave's Before save called and sets ivar to true
3. After create called and updates the record
4. Autosave's Before save gets called a second time and is no longer true
5. We try to set the parent id on the children, but the ivar is false.

If we set the ivar only once, then it is unaffected in subsequent callbacks and we will only take the initial value.

In my case I had been updating a model in an after create with some computations, which is when I noticed my children models weren't associating anymore.

I had hoped to discuss the implications of such a change like this, as I'm not fully aware of what it could imply to other parts of the codebase. It is also interesting to note that updating in most other callbacks create an infinite loop of saving, this one and any after it (e.g. after commit) are the only ones in which updating is possible.

Reply:
Oh also! There is a bug test case linked in the issue that reproduces the issue ☺️ maybe that will help too!

Reply:
Ok. This change now makes sense to me.